ext/drstatecmp/drstatecmp.c

```diff
@@ -0,0 +1,572 @@
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#include "dr_defines.h"
+#include "dr_api.h"
+#include "drmgr.h"
+#include "drstatecmp.h"
+#include <string.h>
+#include <stdint.h>
+
+#ifdef DEBUG
+#    define ASSERT(x, msg) DR_ASSERT_MSG(x, msg)
+#else
+#    define ASSERT(x, msg)
+#endif
+
+typedef struct {
+    bool side_effect_free; /* Denotes whether the current bb has side-effects. */
+    bool orig_bb_mode; /* True when processing the instructions of the original version of
+                          a side-effect-free bb. Becomes false when starting processing
+                           the copies of these instructions*/
+} drstatecmp_user_data_t;
+
+typedef struct {
+    dr_mcontext_t saved_state_for_restore; /* Last saved machine state for restoration. */
+    dr_mcontext_t saved_state_for_cmp;     /* Last saved machine state for comparison. */
+} drstatecmp_saved_states;
+
+static int tls_idx = -1; /* For thread local storage info. */
+
+/* Label types. */
+typedef enum {
+    DRSTATECMP_LABEL_TERM,    /* Denotes the terminator of the original bb. */
+    DRSTATECMP_LABEL_ORIG_BB, /* Denotes the original bb. */
+    DRSTATECMP_LABEL_COPY_BB, /* Denotes the bb copy. */
+    DRSTATECMP_LABEL_COUNT,
+} drstatecmp_label_t;
+
+/* Reserve space for the label values. */
+static ptr_uint_t label_base;
+static void
+drstatecmp_label_init(void)
+{
+    label_base = drmgr_reserve_note_range(DRSTATECMP_LABEL_COUNT);
+    ASSERT(label_base != DRMGR_NOTE_NONE, "failed to reserve note space");
+}
+
+/* Get label values. */
+static inline ptr_int_t
+get_label_val(drstatecmp_label_t label_type)
+{
+    return (ptr_int_t)(label_base + label_type);
+}
+
+/* Compare instr label against a label_type. Returns true if they match. */
+static inline bool
+match_label_val(instr_t *instr, drstatecmp_label_t label_type)
+{
+    return instr_get_note(instr) == (void *)get_label_val(label_type);
+}
+
+/* Create and insert bb labels. */
+static instr_t *
+drstatecmp_insert_labels(void *drcontext, instrlist_t *ilist, instr_t *where,
+                         drstatecmp_label_t label_type, bool preinsert)
+{
+    instr_t *label = INSTR_CREATE_label(drcontext);
+    instr_set_meta(label);
+    instr_set_note(label, (void *)get_label_val(label_type));
+    if (preinsert)
+        instrlist_meta_preinsert(ilist, where, label);
+    else
+        instrlist_meta_postinsert(ilist, where, label);
+    return label;
+}
+
+/* Returns whether or not instr may have side effects. */
+static bool
+drstatecmp_may_have_side_effects_instr(instr_t *instr)
+{
+    /* Instructions with side effects include instructions that write to memory,
+     * interrupts, and syscalls. To avoid inter-procedural analysis, all function calls
+     * are conservatively assumed to have side effects.
```

Comment:
The function call comment can be removed.

Reply:
Removed.